pubtools/_pulp/tasks/copy_repo.py

```diff
@@ -0,0 +1,324 @@
+            ]
+        )
+        if container_repo_ids:
+            self.fail(
+                "Container image repo(s) provided, not supported: %s"
+                % ", ".join(sorted(container_repo_ids))
+            )
+
+        return src_repo, dest_repo
+
+    def log_copy(self, copied_repo):
+        # Given a repo which has been copied, log some messages
+        # summarizing the copied unit(s)
+        content_types = {}
+
+        for task in copied_repo.tasks:
+            for unit in task.units:
+                type_id = unit.content_type_id
+                content_types[type_id] = content_types.get(type_id, 0) + 1
+
+        task_ids = ", ".join(sorted([t.id for t in copied_repo.tasks]))
+        repo_id = copied_repo.repo.id
+        if not content_types:
+            LOG.warning("%s: no content copied, tasks: %s", repo_id, task_ids)
+        else:
+            copied_types = []
+            for key in sorted(content_types.keys()):
+                copied_types.append("%s %s(s)" % (content_types[key], key))
+            copied_types = ", ".join(copied_types)
+
+            LOG.info("%s: removed %s, tasks: %s", repo_id, copied_types, task_ids)
+
+        return copied_repo
+
+    @step("Copy content")
+    def copy_content(self, src_repo, dest_repo):
+        futures = []
+        content_types = self.content_type or ["None"]
+        for t in content_types:
+            crit = None
+            if t == "iso":
+                crit = Criteria.with_unit_type(FileUnit)
+            if t in ("rpm", "srpm"):
+                crit = Criteria.with_unit_type(RpmUnit)
+            if t == "modulemd":
+                crit = Criteria.with_unit_type(ModulemdUnit)
+
+            f = self.pulp_client.copy_content(src_repo, dest_repo, criteria=crit)
+            f = f_map(f, partial(RepoCopy, repo=dest_repo))
+            f = f_map(f, self.log_copy)
+            futures.append(f)
+
+        return futures
+
+    @step("Record push items")
+    def record_copies(self, copied_repos_fs):
+        return [f_flat_map(f, self.record_copied_repo) for f in copied_repos_fs]
+
+    def record_copied_repo(self, copied_repo):
+        push_items = []
+        for task in copied_repo.tasks:
+            push_items.extend(self.push_items_for_task(task))
+        return self.collector.update_push_items(push_items)
+
+    def push_items_for_task(self, task):
+        out = []
+        for unit in task.units:
+            push_item = self.push_item_for_unit(unit)
+            if push_item:
+                out.append(push_item)
+        return out
+
+    def push_item_for_unit(self, unit):
+        for unit_type, fn in [
+            (ModulemdUnit, self.push_item_for_modulemd),
+            (RpmUnit, self.push_item_for_rpm),
+            (FileUnit, self.push_item_for_file),
+        ]:
+            if isinstance(unit, unit_type):
+                return fn(unit)
+
+    def push_item_for_modulemd(self, unit):
```

Comment:
If these methods have been copy-pasted, can you please find a way to share them instead of duplicating them?